src/html/link.ml

```diff
@@ -73,6 +73,7 @@ let href ~resolve { Url.Anchor. page; anchor; kind } =
       @ target_from_common_ancestor
     in
     let page = String.concat "/" relative_target in
+    let page = if kind="page" then page ^ ".html" else page in
```

Comment:
That's not the right place, you should fix `Path.for_linking` above. Looking at it, the implementation should probably merge code with `as_filename` as much as possible...